### PR TITLE
Cross-dataset: Layer-wise LR Decay (LLRD) — differentiated learning rates per layer (CODE CHANGE)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Standard training assigns the same learning rate to all transformer layers. In transfer learning (ViT fine-tuning), Layer-wise LR Decay (LLRD) consistently helps: earlier (geometry encoder) layers learn slower than later (physics decoder) layers. Our model has similar structure — early transformer layers learn geometric representations while later layers specialize to physics fields. We hypothesize that LLRD will stabilize early-layer geometry representations while allowing later layers to adapt more aggressively, improving generalization across all three datasets.

Reference: https://arxiv.org/abs/1905.05583 (ULMFiT), standard in ViT fine-tuning (BEiT, DeiT).

## Instructions

**CODE CHANGE REQUIRED** — add `--llrd-decay` argument and LLRD optimizer logic to `train.py`.

### Step 1: Add the argument

```python
parser.add_argument('--llrd-decay', type=float, default=1.0,
                    help='Layer-wise LR decay factor. 1.0 = disabled. 0.75 means each earlier layer gets 0.75x the LR of the next layer.')
```

### Step 2: Replace the optimizer construction block

After `args.lr` and `args.weight_decay` are parsed, wrap the optimizer creation:

```python
if args.llrd_decay < 1.0:
    n_layers = args.model_layers
    param_groups = []
    for i, layer in enumerate(model.transformer_layers):
        layer_lr = args.lr * (args.llrd_decay ** (n_layers - 1 - i))
        param_groups.append({"params": layer.parameters(), "lr": layer_lr})
    other_params = [p for n, p in model.named_parameters() if 'transformer_layers' not in n]
    param_groups.append({"params": other_params, "lr": args.lr})
    optimizer = AdamW(param_groups, weight_decay=args.weight_decay)
else:
    # existing optimizer construction (unchanged)
    ...
```

### Step 3: Run all 8 experiments

Use `--wandb_group llrd-cross` for all runs.

**DrivAerML — LLRD 0.75:**
```bash
python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --llrd-decay 0.75 --wandb_group llrd-cross
```

**DrivAerML — LLRD 0.85:**
```bash
python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --llrd-decay 0.85 --wandb_group llrd-cross
```

**AirfRANS — LLRD 0.75:**
```bash
python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --llrd-decay 0.75 --wandb_group llrd-cross
```

**AirfRANS — LLRD 0.85:**
```bash
python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --llrd-decay 0.85 --wandb_group llrd-cross
```

**TandemFoil — LLRD 0.75:**
```bash
python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --llrd-decay 0.75 --wandb_group llrd-cross
```

**TandemFoil — LLRD 0.85:**
```bash
python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --llrd-decay 0.85 --wandb_group llrd-cross
```

**TandemFoil — LLRD 0.65:**
```bash
python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --llrd-decay 0.65 --wandb_group llrd-cross
```

**DrivAerML — LLRD 0.65:**
```bash
python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --llrd-decay 0.65 --wandb_group llrd-cross
```

## Baseline

Current best metrics:
- **TandemFoil**: val/loss = 30.10 (epoch 157, PR #2810, W&B run v6amjkh7)
- **AirfRANS**: val/loss = 0.001236 (epoch 349, PR #2828, W&B run libpwryz)
- **DrivAerML**: val/loss = 4.619% (epoch 256, PR #2691, W&B run k8qtsxxz)

Reproduce commands:
```bash
# TandemFoil baseline
cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999

# AirfRANS baseline
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999

# DrivAerML baseline
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
Closing to narrow the queue for the final ICML benchmark sprint. Current priority is benchmark-specific optimization and benchmark-faithful reporting for DrivAerML, TandemFoil paper benchmark, AirfRANS full Surf/Vol, and the Tandem parity anchor. If this idea still matters later, we can reopen it in a benchmark-specific form.
